Update and expand theory sections (#3114)
ROCm/rocm-systems: projects/hip/docs/understand/hardware_implementation.rst

Reviewer: ```suggestion
This topic describes the hardware architecture of AMD GPUs supported by HIP,
```

Overall GPU architecture
========================

AMD GPUs consist of interconnected blocks of digital circuits that work
together to execute complex parallel computing tasks. Unlike central
processing units (CPUs), which dedicate significant silicon area to
instruction flow control, branch prediction, and complex caching hierarchies,
GPUs allocate the majority of their die area to arithmetic pipelines. This
design choice enables extreme throughput density for data-parallel workloads.
The architecture is organized hierarchically to enable massive parallelism

> while managing resources efficiently.

Reviewer: ```suggestion
while efficiently managing resources.
```

1. **Shader engines (SE)**: Top-level organizational units containing multiple
   shader arrays and shared resources
2. **Shader arrays**: Groups of compute units (CUs) sharing instruction and
   scalar caches
3. **Compute units (CU)**: Basic execution units containing the arithmetic
   logic units (ALUs) and registers for thread execution

.. figure:: ../data/understand/hardware_implementation/selayout.png
   :align: center
   :alt: Diagram showing the hierarchical organization of compute units
         grouped into shader engines on AMD GPUs
   :width: 800

   Hierarchical organization of compute units into shader engines

This hierarchical design allows different GPU configurations using the same
underlying architecture.

Shader engine components
========================

Shader engines group multiple compute units together with shared resources
that improve efficiency and reduce redundancy. Each shader engine contains

> several key components shared across its compute units.

Reviewer: ```suggestion
Shader engines group multiple compute units together, sharing resources to improve efficiency and reduce redundancy. Each shader engine contains
several key components shared across its compute units.
```

L1 instruction cache (L1I)
--------------------------

> The L1 instruction cache (L1I) is a read-only cache shared between multiple
> CUs in a shader array. Like the sL1D, it is backed by the L2 cache and doesn't
> use the "hit-on-miss" approach. The L1I stores kernel instructions fetched by
> the compute units, reducing instruction fetch latency and L2 cache pressure.

Reviewer: ```suggestion
The L1 instruction cache (L1I) is a read-only cache shared between multiple
CUs in a shader array. Like the sL1D, it's backed by the L2 cache and doesn't
use the "hit-on-miss" approach. The L1I stores kernel instructions fetched by
the compute units, reducing instruction fetch latency and L2 cache pressure.
```

.. _compute_unit:

Compute unit architecture
=========================

The compute unit (CU) is the fundamental execution block of AMD GPUs, serving
as the atomic building block for massive parallelism. Each CU is responsible
for executing kernels through its various specialized components and
pipelines. Data flows into these pipelines, undergoes arithmetic
transformation, and exits as results, with the goal of maximizing the number

> of such transformations per clock cycle.

Reviewer: ```suggestion
transformation, and exits as results, to maximize the number
of such transformations per clock cycle.
```

   Internal architecture of an AMD CDNA compute unit

.. _wave-scheduling:

Sequencer and scheduling
------------------------

The instruction sequencer (SQ) serves as the control center of each compute
unit, managing instruction flow through the execution pipelines. The sequencer
maintains wavefront state and coordinates instruction execution across
different functional units.

**Wavefront organization**: The sequencer organizes active wavefronts into
four pools, each containing slots for up to ten wavefronts (eight on the CDNA2

> MI200 series). Each slot includes:

Reviewer: ```suggestion
MI200 Series). Each slot includes:
```

**Instruction fetching**: The fetch arbiter selects one wavefront per cycle to
fetch instructions from memory, prioritizing the oldest wavefronts. Each CU
can fetch up to 32 bytes (4-8 instructions) per cycle.

**Instruction issuing**: The issue arbiter determines which instructions
execute each cycle, selecting wavefronts from one pool per cycle in
round-robin fashion. The arbiter can issue multiple instructions per cycle to
different execution units, with a theoretical maximum of five instructions per
cycle:

* One VALU instruction
* One vector memory operation
* One SALU and/or scalar memory operation
* One LDS operation
* One branch operation

Instructions always issue at wavefront granularity, with all threads in the
wavefront executing the same instruction in lockstep. The hardware can perform
single-cycle context switching between wavefronts with zero overhead, as all
wavefront contexts remain resident on the CU. This enables efficient latency
hiding, allowing the CU to switch to another wavefront immediately when the
current wavefront encounters a stall condition such as a memory access.

Execution pipelines
-------------------

> Each compute unit contains multiple specialized execution pipelines that

Reviewer 2: ```suggestion
Each CU contains multiple specialized execution pipelines that
```

flow, not execution mask updates for thread divergence, which are handled
through predication.

.. _sfu:

Special function unit (SFU)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The special function units accelerate certain arithmetic operations that are
too complex and/or costly to implement purely within the standard vector ALUs.

SFUs are responsible for executing transcendental and reciprocal mathematical
functions, operations such as ``exp``, ``log``, ``sin``, ``cos``, ``rcp``
(reciprocal), and ``rsqrt`` (reciprocal square root). These are heavily used
in scientific, physics, and machine learning workloads, particularly in
activation functions such as GELU, sigmoid, and/or softmax.

> Each compute unit includes a set of specialized pipelines and/or transcendental

Reviewer 2: ```suggestion
Each CU includes a set of specialized pipelines and/or transcendental
```

* ``v_log_f32`` - compute natural logarithm
* ``v_sin_f32``, ``v_cos_f32`` - compute sine and/or cosine
* ``v_rsq_f32``, ``v_rcp_f32`` - compute reciprocal and/or reciprocal square
  root

In CDNA3-based GPUs (such as MI300), SFU throughput and latency have been
tuned for deep learning primitives. For instance, exponentiation (``exp``) and
logarithm (``log``) functions are now pipelined to complete in a few cycles
per lane, allowing vectorized activation functions in large-scale matrix
workloads to execute without significant stalls.

For programmers targeting ROCm and/or HIP, these SFU-accelerated operations
are typically accessed through math intrinsics such as ``__expf``, ``__logf``,
and/or ``__sinf``, which the compiler lowers to the corresponding AMDGPU ISA
instructions at compile time.

.. _lsu:

Load/store unit (LSU)
^^^^^^^^^^^^^^^^^^^^^

The load/store units handle the transfer of data between the compute units and
the GPU's memory subsystems. They are responsible for issuing, tracking, and
retiring memory operations, including loads from and stores to global memory,
local shared memory, and caches, for thousands of concurrent threads.

> Each compute unit includes a set of LSUs tightly integrated with its vector

Reviewer 2: ```suggestion
Each CU includes a set of LSUs tightly integrated with its vector
```

multiply-add instructions, each MFMA instruction processes an entire matrix
fragment, drastically reducing power per operation and increasing overall
throughput. The MFMA units implement a mini-systolic array design that
efficiently processes matrix tiles.

An example MFMA instruction from the AMDGPU ISA is:

.. code-block:: amdgpu

   v_mfma_f32_16x16x4f16 v[0:15], v[16:31], v[32:47], v[0:15]

This instruction performs a matrix multiplication and accumulation
:math:`\pmb{D}=\pmb{A} \times \pmb{B} + \pmb{C}`, where the fragments
:math:`\pmb{A}`, :math:`\pmb{B}`, and :math:`\pmb{C}` are stored in VGPRs.
The suffix ``16x16x4f16`` indicates a tile size of :math:`16 \times 16`, with
an inner dimension of :math:`4`, operating on half-precision (FP16) inputs and
accumulating into 32-bit floating-point outputs.

Programmers can access MFMA functionality at multiple levels: through
optimized libraries, compiler intrinsics, and/or inline assembly, providing
flexibility for different use cases.

The MFMA units use both standard VGPRs and additional accumulation VGPRs
(AGPRs) on supported architectures, providing up to 512 KiB of combined
register storage per CU.

.. _dme:

> Data movement engine (CDNA3+)

Reviewer: What does CDNA3+ mean?
Perhaps CDNA3/CDNA4?

Author: Well, it will probably also land in future generations. But we can change it to CDNA3/CDNA4 for now.

Author: Fixed.

   Local data share organization and SIMD connections

> **Organization**: The LDS contains 32 (CDNA, CDNA2, and CDNA3) or 64 (CDNA4+ and

Reviewer: what is CDNA4+?

Reviewer: We haven't released that yet, have we?

Author: Fixed.

RDNA2+) banks, each 4-bytes wide, providing 128 (CDNA, CDNA2, and CDNA3) or 256
(CDNA4+ and RDNA2+) bytes per cycle total bandwidth. Banks can be accessed
independently each cycle for reads, writes, and/or atomic operations. The SIMDs
connect to the LDS in pairs, with each pair sharing a 64-byte bidirectional

> port.

Reviewer: This sentence seems off:
Banks can be accessed independently each cycle?

Author: Rephrased a bit to make the intention clearer.
